pkg/skaffold/initializer/analyze.go

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2020 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package initializer
+
+import (
+	"fmt"
+	"path/filepath"
+	"sort"
+
+	"github.com/karrick/godirwalk"
+	"github.com/sirupsen/logrus"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/initializer/kubectl"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+)
+
+type analysis struct {
+	kubectlAnalyzer  *KubectlAnalyzer
+	skaffoldAnalyzer *SkaffoldConfigAnalyzer
+	builderAnalyzer  *BuilderAnalyzer
+}
+
+// Analyzer is a generic Visitor that is called on every file in the directory
+// It can manage state and react to walking events assuming a bread first search
+type Analyzer interface {
+	EnterDir(dir string)
+	AnalyzeFile(file string) error
+	ExitDir(dir string)
+}
+
+type DirectoryAnalyzer struct {
+	currentDir string
+}
+
+func (a *DirectoryAnalyzer) AnalyzeFile(filePath string) error {
+	return nil
+}
+
+func (a *DirectoryAnalyzer) EnterDir(dir string) {
+	a.currentDir = dir
+}
+func (a *DirectoryAnalyzer) ExitDir(dir string) {
+	//pass
+}
+
+type KubectlAnalyzer struct {
+	DirectoryAnalyzer
+	kubernetesManifests []string
+}
+
+func (a *KubectlAnalyzer) AnalyzeFile(filePath string) error {
+	isSkaffoldConfig := IsSkaffoldConfig(filePath)
+	isKubernetesManifest := kubectl.IsKubernetesManifest(filePath)
+
+	if isKubernetesManifest && !isSkaffoldConfig {
+		a.kubernetesManifests = append(a.kubernetesManifests, filePath)
+	}
+	return nil
+}
+
+type SkaffoldConfigAnalyzer struct {
+	DirectoryAnalyzer
+	force bool
+}
+
+func (a *SkaffoldConfigAnalyzer) AnalyzeFile(filePath string) error {
+	isSkaffoldConfig := IsSkaffoldConfig(filePath)
+	if isSkaffoldConfig {
```

Comment:
nit: what about inverting the test? guard vs nesting

Reply:
great point!